### PR TITLE
Issue 21826: Add prototype code for creating private_key_jwt client authentication

### PR DESCRIPTION
**Pull request by ayoho**
- Adds logic to the OIDC token endpoint request code to add `private_key_jwt` related parameters when using that token endpoint auth method
- Adds a large part of the logic to the `PrivateKeyJwtAuthMethod` class to create the JWT for client authentication
    - Note: Still missing logic to obtain the key itself using the keyAliasName
    
  For #21826

### CODE REVIEW

**Comment by ayoho** [comment]
#build
#spawn.fullfat.buckets=com.ibm.ws.security.oauth_fat,com.ibm.ws.security.oidc.client_fat.1,com.ibm.ws.security.oidc.client_fat.2,com.ibm.ws.security.oidc.client_fat.backchannelLogout,com.ibm.ws.security.oidc.client_fat.claimPropagation,com.ibm.ws.security.oidc.client_fat.jaxrs,com.ibm.ws.security.oidc.client_fat.spnego,com.ibm.ws.security.oidc.server_fat.backchannelLogout,com.ibm.ws.security.oidc.server_fat.jaxrs.config.commonTest,com.ibm.ws.security.oidc.server_fat.jaxrs.config.noOP,com.ibm.ws.security.oidc.server_fat.jaxrs.config.oauth,com.ibm.ws.security.oidc.server_fat.jaxrs.config.oidc,com.ibm.ws.security.oidc.server_fat.oidc,com.ibm.ws.security.social_fat,com.ibm.ws.security.social_fat.LibertyOP.1,com.ibm.ws.security.social_fat.LibertyOP.2,com.ibm.ws.security.social_fat.LibertyOP.backchannelLogout,com.ibm.ws.security.social_fat.LibertyOP.claimPropagation,com.ibm.ws.security.social_fat.OpenShift,com.ibm.ws.security.social_fat.delegated,com.ibm.ws.security.social_fat.multiProvider,com.ibm.ws.security.social_fat.okdServiceLogin,io.openliberty.security.jakartasec.3.0.internal_fat,io.openliberty.security.jakartasec.3.0.internal_fat.config.1,io.openliberty.security.jakartasec.3.0.internal_fat.config.2,io.openliberty.security.jakartasec.3.0.internal_fat.logout,io.openliberty.security.jakartasec.3.0.internal_fat.refresh

**Comment by posmith** [comment]
L2 message review completed

File: dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/exceptions/PrivateKeyJwtAuthException.java [hunk 1]
```diff
@@ -0,0 +1,34 @@
+/*******************************************************************************
+ * Copyright (c) 2022 IBM Corporation and others.
```
Comment (jimmy1wu) [inline dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/exceptions/PrivateKeyJwtAuthException.java:2]:
```suggestion
 * Copyright (c) 2023 IBM Corporation and others.
```

File: dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/exceptions/PrivateKeyJwtAuthMissingKeyException.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
+/*******************************************************************************
+ * Copyright (c) 2022 IBM Corporation and others.
```
Comment (jimmy1wu) [inline dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/exceptions/PrivateKeyJwtAuthMissingKeyException.java:2]:
```suggestion
 * Copyright (c) 2023 IBM Corporation and others.
```

File: dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/token/TokenRequestor.java [hunk 3]
```diff
@@ -99,7 +106,7 @@ private List<NameValuePair> getBasicParams() {
         return params;
     }
 
-    void addPrivateKeyJwtParameters(List<NameValuePair> params) {
+    void addPrivateKeyJwtParameters(List<NameValuePair> params) throws TokenRequestException {
```
Comment (jimmy1wu) [inline dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/token/TokenRequestor.java:109]: **code enhancement opportunities for next pr:**

this method can probably be moved to:
dev/com.ibm.ws.security.openidconnect.clients.common/src/com/ibm/ws/security/openidconnect/common/token/auth/PrivateKeyJwtAuthMethod.java

we can pass in the params via `customParams` in the builder.

`PrivateKeyJwtAuthMethod` will end up being something like

```java
    @Override
    public void setAuthMethodSpecificSettings(Builder tokenRequestBuilder) {
        tokenRequestBuilder.clientAssertionSigningAlgorithm(clientConfig.getTokenEndpointAuthSigningAlgorithm());
        Key clientAssertionSigningKey = getKeyForPrivateKeyJwtClientAssertion();
        tokenRequestBuilder.clientAssertionSigningKey(clientAssertionSigningKey);

        vvvvvvv
        Map<String, String> clientAssertionParams = getClientAssertionParams();
        tokenRequestBuilder.customParams(clientAssertionParams);
        ^^^^^^^
    }
    
    private Map<String, String> getClientAssertionParams() {
        Map<String, String> clientAssertionParams = new HashMap<>();
        clientAssertionParams.add(TokenConstants.CLIENT_ASSERTION_TYPE, TokenConstants.CLIENT_ASSERTION_TYPE_JWT_BEARER);
        String clientAssertionJwt = buildClientAssertionJwt();
        clientAssertionParams.add(TokenConstants.CLIENT_ASSERTION, clientAssertionJwt);
        return clientAssertionParams;
    }
    ...
```

perhaps the `createPrivateKeyJwt` method from:
dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/token/auth/PrivateKeyJwtAuthMethod.java can be moved to here as well, so we don't end up with 2 files named `PrivateKeyJwtAuthMethod`.

also, if we create a `PostAuthMethod` and `ClientSecretPostAuthMethod` which both extends from your new `TokenEndpointAuthMethod`, then we can get rid of this chunk entirely:
https://github.com/OpenLiberty/open-liberty/blob/e46a4d1210048ec9bdb2967d84f6783ebcc69068/dev/io.openliberty.security.oidcclientcore.internal/src/io/openliberty/security/oidcclientcore/token/TokenRequestor.java#L100-L105

`PostAuthMethod` and `ClientSecretPostAuthMethod` can both just pass in the `clientId` and `clientSecret` using the same `customParams` method i presented above.

**Review by jimmy1wu — APPROVED** [review]

File: dev/io.openliberty.security.oidcclientcore.internal/resources/io/openliberty/security/oidcclientcore/internal/resources/OidcClientCoreMessages.nlsprops [hunk 4]
```diff
@@ -139,3 +139,10 @@ TOKEN_RESPONSE_MISSING_PARAMETER=CWWKS2429E: The token response from the OpenID
 TOKEN_RESPONSE_MISSING_PARAMETER.explanation=Either the OpenID Connect provider did not return a required parameter or failed to process the token request.
 TOKEN_RESPONSE_MISSING_PARAMETER.useraction=Verify that the configured token endpoint is correct and capable of responding to OAuth token requests.
 
+PRIVATE_KEY_JWT_AUTH_ERROR=CWWKS2430E: The [{0}] OpenID Connect client failed to build a JSON Web Token to use for client authentication. {1}
+PRIVATE_KEY_JWT_AUTH_ERROR.explanation=The OpenID Connect client is configured to use JSON Web Tokens for client authentication, but an error occurred while creating the token.
+PRIVATE_KEY_JWT_AUTH_ERROR.useraction=For more information, see the error in the message.
+
+PRIVATE_KEY_JWT_MISSING_SIGNING_KEY=CWWKS2431E: The [{0}] OpenID Connect client is missing the key that is needed to sign the token that is used for client authentication.
```
Comment (helyarp) [inline dev/io.openliberty.security.oidcclientcore.internal/resources/io/openliberty/security/oidcclientcore/internal/resources/OidcClientCoreMessages.nlsprops:146]: Remove "that is used" unless the removal will make the message technically incorrect. (Change "to sign the token that is used for client authentication" to "to sign the token for client authentication".)

File: dev/io.openliberty.security.oidcclientcore.internal/resources/io/openliberty/security/oidcclientcore/internal/resources/OidcClientCoreMessages.nlsprops [hunk 5]
```diff
@@ -139,3 +139,10 @@ TOKEN_RESPONSE_MISSING_PARAMETER=CWWKS2429E: The token response from the OpenID
...
+PRIVATE_KEY_JWT_MISSING_SIGNING_KEY.explanation=The OpenID Connect client configuration is missing data, or a key cannot be found with the key alias name in the SSL configuration that is specified in the OpenID Connect client.
```
Comment (helyarp) [inline dev/io.openliberty.security.oidcclientcore.internal/resources/io/openliberty/security/oidcclientcore/internal/resources/OidcClientCoreMessages.nlsprops:147]: Change "that is specified in" to "for"

File: dev/io.openliberty.security.oidcclientcore.internal/resources/io/openliberty/security/oidcclientcore/internal/resources/OidcClientCoreMessages.nlsprops [hunk 6]
```diff
@@ -139,3 +139,10 @@ TOKEN_RESPONSE_MISSING_PARAMETER=CWWKS2429E: The token response from the OpenID
...
+PRIVATE_KEY_JWT_MISSING_SIGNING_KEY.useraction=Ensure that the OpenID Connect client has a key alias name and SSL reference configured. Verify that the keystore that is referenced by the SSL configuration contains a key whose alias matches the key alias name in the OpenID Connect client configuration.
```
Comment (helyarp) [inline dev/io.openliberty.security.oidcclientcore.internal/resources/io/openliberty/security/oidcclientcore/internal/resources/OidcClientCoreMessages.nlsprops:148]: Change "that is referenced by the SSL configuration contains" to "that the SSL configuration references contains"

**Review by helyarp — COMMENTED** [review]
Suggested changes

**Pull request merged**